File: stepper_stabilizer/controller.py
"""PID control logic, kept independent from Raspberry Pi hardware."""

from dataclasses import dataclass
from enum import Enum
import math
from typing import Optional


class Direction(Enum):
    """Logical level sent to both DRV8825 DIR inputs."""

    REVERSE = 0
    FORWARD = 1


@dataclass(frozen=True)
class Correction:
    """One step burst requested by a PID update."""

    direction: Direction
    steps: int
    low_delay_seconds: float


@dataclass(frozen=True)
class PidConfig:
    """PID values and movement limits carried over from the ESP32 sketch."""

    kp: float = 0.80
    ki: float = 0.05
    kd: float = 0.10
    integral_limit: float = 0.50
    output_limit: float = 1.0
    dead_band_g: float = 0.15
    fast_low_delay_seconds: float = 0.0025
    slow_low_delay_seconds: float = 0.015

    def __post_init__(self) -> None:
        if self.integral_limit < 0:
            raise ValueError("integral_limit must not be negative")
        if self.output_limit <= 0:
            raise ValueError("output_limit must be positive")
        if self.dead_band_g < 0:
            raise ValueError("dead_band_g must not be negative")
        if not 0 < self.fast_low_delay_seconds <= self.slow_low_delay_seconds:
            raise ValueError(
                "step delays must satisfy 0 < fast_low_delay <= slow_low_delay"
            )


class PidController:
    """Convert signed X-axis acceleration into synchronized motor corrections."""

    def __init__(self, config: Optional[PidConfig] = None) -> None:
        self.config = config or PidConfig()
        self._previous_error: Optional[float] = None
        self._integral_error = 0.0

    def update(
        self, acceleration_x_g: float, elapsed_seconds: float
    ) -> Optional[Correction]:
        """Process one sample.

        The target is 0 g, so ``acceleration_x_g`` is the signed error. The
        first valid sample initializes the derivative term without moving.
        """

        if not math.isfinite(acceleration_x_g):
            return None

        if self._previous_error is None:
            self._previous_error = acceleration_x_g
            return None

        if not math.isfinite(elapsed_seconds) or elapsed_seconds <= 0:
            return None

        error = acceleration_x_g
        proportional = self.config.kp * error

        if abs(error) < self.config.dead_band_g:
            self._integral_error = 0.0
        else:
            self._integral_error = _clamp(
                self._integral_error + error * elapsed_seconds,
                -self.config.integral_limit,
                self.config.integral_limit,
            )

        integral = self.config.ki * self._integral_error
        derivative = (
            self.config.kd
            * (error - self._previous_error)
            / elapsed_seconds
        )
        output = _clamp(
            proportional + integral + derivative,
            -self.config.output_limit,
            self.config.output_limit,
        )

        self._previous_error = error

        if abs(output) < self.config.dead_band_g:
            return None

        direction = Direction.FORWARD if output > 0 else Direction.REVERSE
        strength = _clamp(
            abs(output) / self.config.output_limit,
            0.0,
            1.0,
        )
        low_delay = (
            self.config.slow_low_delay_seconds
            - strength
            * (
                self.config.slow_low_delay_seconds
                - self.config.fast_low_delay_seconds
            )
        )

        return Correction(
            direction=direction,
            steps=_steps_for_error(error),
            low_delay_seconds=low_delay,
        )


def _steps_for_error(error: float) -> int:
    absolute_error = abs(error)
    if absolute_error < 0.2:
        return 2
    if absolute_error < 0.4:
        return 4
    if absolute_error < 0.6:
        return 8
    return 20


def _clamp(value: float, minimum: float, maximum: float) -> float:
    return max(minimum, min(value, maximum))

File: src/controller.rs
use std::time::Duration;

/// Direction level sent to both DRV8825 DIR inputs.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum Direction {
    Reverse,
    Forward,
}

/// A motor movement requested by one PID update.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct Correction {
    pub direction: Direction,
    pub steps: u32,
    pub low_delay: Duration,
}

/// PID gains and movement limits carried over from the ESP32 sketch.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct PidConfig {
    pub kp: f32,
    pub ki: f32,
    pub kd: f32,
    pub integral_limit: f32,
    pub output_limit: f32,
    pub dead_band_g: f32,
    pub fast_low_delay: Duration,
    pub slow_low_delay: Duration,
}

impl Default for PidConfig {
    fn default() -> Self {
        Self {
            kp: 0.80,
            ki: 0.05,
            kd: 0.10,
            integral_limit: 0.50,
            output_limit: 1.0,
            dead_band_g: 0.10,
            fast_low_delay: Duration::from_micros(2_500),
            slow_low_delay: Duration::from_micros(15_000),
        }
    }
}

/// Stateful PID controller that converts acceleration error into step bursts.
#[derive(Debug)]
pub struct PidController {
    config: PidConfig,
    previous_error: Option<f32>,
    integral_error: f32,
}

impl PidController {
    pub fn new(config: PidConfig) -> Self {
        Self {
            config,
            previous_error: None,
            integral_error: 0.0,
        }
    }

    /// Updates the controller.
    ///
    /// The target is 0 g, so `acceleration_x_g` is also the signed error.
    /// The first valid sample initializes the derivative term without moving.
    pub fn update(&mut self, acceleration_x_g: f32, elapsed: Duration) -> Option<Correction> {
        if !acceleration_x_g.is_finite() {
            return None;
        }

        let Some(previous_error) = self.previous_error else {
            self.previous_error = Some(acceleration_x_g);
            return None;
        };

        let elapsed_seconds = elapsed.as_secs_f32();
        if !elapsed_seconds.is_finite() || elapsed_seconds <= 0.0 {
            return None;
        }

        let error = acceleration_x_g;
        let proportional = self.config.kp * error;

        if error.abs() < self.config.dead_band_g {
            self.integral_error = 0.0;
        } else {
            self.integral_error = (self.integral_error + error * elapsed_seconds)
                .clamp(-self.config.integral_limit, self.config.integral_limit);
        }

        let integral = self.config.ki * self.integral_error;
        let derivative = self.config.kd * (error - previous_error) / elapsed_seconds;
        let output = (proportional + integral + derivative)
            .clamp(-self.config.output_limit, self.config.output_limit);

        self.previous_error = Some(error);

        if output.abs() < self.config.dead_band_g {
            return None;
        }

        let direction = if output > 0.0 {
            Direction::Forward
        } else {
            Direction::Reverse
        };
        let strength = (output.abs() / self.config.output_limit).clamp(0.0, 1.0);
        let fast_us = self.config.fast_low_delay.as_micros() as f32;
        let slow_us = self.config.slow_low_delay.as_micros() as f32;
        let low_delay_us = slow_us - strength * (slow_us - fast_us);

        Some(Correction {
            direction,
            steps: steps_for_error(error),
            low_delay: Duration::from_micros(low_delay_us as u64),
        })
    }
}

fn steps_for_error(error: f32) -> u32 {
    match error.abs() {
        value if value < 0.2 => 2,
        value if value < 0.4 => 4,
        value if value < 0.6 => 8,
        _ => 20,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const SAMPLE_TIME: Duration = Duration::from_millis(50);

    fn initialized_controller(initial_error: f32) -> PidController {
        let mut controller = PidController::new(PidConfig::default());
        assert_eq!(
            controller.update(initial_error, SAMPLE_TIME),
            None,
            "the first sample must only initialize the derivative term"
        );
        controller
    }

    #[test]
    fn ignores_invalid_samples() {
        let mut controller = PidController::new(PidConfig::default());
        assert_eq!(controller.update(f32::NAN, SAMPLE_TIME), None);
        assert_eq!(controller.update(f32::INFINITY, SAMPLE_TIME), None);
        assert_eq!(controller.update(0.5, SAMPLE_TIME), None);
    }

    #[test]
    fn does_not_move_inside_dead_band() {
        let mut controller = initialized_controller(0.05);
        assert_eq!(controller.update(0.05, SAMPLE_TIME), None);
    }

    #[test]
    fn positive_and_negative_errors_select_opposite_directions() {
        let mut positive = initialized_controller(0.3);
        let mut negative = initialized_controller(-0.3);

        assert_eq!(
            positive.update(0.3, SAMPLE_TIME).unwrap().direction,
            Direction::Forward
        );
        assert_eq!(
            negative.update(-0.3, SAMPLE_TIME).unwrap().direction,
            Direction::Reverse
        );
    }

    #[test]
    fn preserves_the_original_step_bands() {
        for (error, expected_steps) in [
            (0.19, 2),
            (0.20, 4),
            (0.39, 4),
            (0.40, 8),
            (0.59, 8),
            (0.60, 20),
        ] {
            let mut controller = initialized_controller(error);
            assert_eq!(
                controller.update(error, SAMPLE_TIME).unwrap().steps,
                expected_steps,
                "unexpected step count for {error} g"
            );
        }
    }

    #[test]
    fn clamps_full_output_to_fastest_delay() {
        let mut controller = initialized_controller(0.0);
        let correction = controller.update(10.0, SAMPLE_TIME).unwrap();
        assert_eq!(correction.low_delay, Duration::from_micros(2_500));
    }

    #[test]
    fn zero_elapsed_time_does_not_change_controller_history() {
        let mut controller = initialized_controller(0.2);
        assert_eq!(controller.update(0.8, Duration::ZERO), None);

        let correction = controller.update(0.2, SAMPLE_TIME).unwrap();
        assert_eq!(correction.steps, 4);
    }
}
